plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/tracing/TracingHiveMetastore.java

```diff
@@ -0,0 +1,712 @@
+    {
+        Span span = startSpan("getFunctions", databaseName, functionName);
+        try (var ignored = scopedSpan(span)) {
+            return delegate.getFunctions(databaseName, functionName);
+        }
+    }
+
+    @Override
+    public void createFunction(String databaseName, String functionName, LanguageFunction function)
+    {
+        Span span = startSpan("createFunction", databaseName)
+                .setAttribute(FUNCTION, functionName);
+        try (var ignored = scopedSpan(span)) {
+            delegate.createFunction(databaseName, functionName, function);
+        }
+    }
+
+    @Override
+    public void replaceFunction(String databaseName, String functionName, LanguageFunction function)
+    {
+        Span span = startSpan("replaceFunction", databaseName)
+                .setAttribute(FUNCTION, functionName);
+        try (var ignored = scopedSpan(span)) {
+            delegate.replaceFunction(databaseName, functionName, function);
+        }
+    }
+
+    @Override
+    public void dropFunction(String databaseName, String functionName, String signatureToken)
+    {
+        Span span = startSpan("dropFunction", databaseName)
+                .setAttribute(FUNCTION, functionName);
+        try (var ignored = scopedSpan(span)) {
+            delegate.dropFunction(databaseName, functionName, signatureToken);
+        }
+    }
+
+    private Span startSpan(String methodName)
+    {
+        return tracer.spanBuilder("Metadata." + methodName)
+                .startSpan();
+    }
+
+    private Span startSpan(String methodName, String schemaName)
+    {
+        return tracer.spanBuilder("Metadata." + methodName)
+                .startSpan()
+                .setAttribute(SCHEMA, schemaName);
+    }
+
+    private Span startSpan(String methodName, Table table)
+    {
+        return startSpan(methodName, table.getDatabaseName(), table.getTableName());
+    }
+
+    private Span startSpan(String methodName, String schemaName, String tableName)
+    {
+        return tracer.spanBuilder("Metadata." + methodName)
+                .startSpan()
+                .setAttribute(SCHEMA, schemaName)
+                .setAttribute(TABLE, tableName);
+    }
+
+    private Span startSpan(String methodName, Table table, String partitionName)
+    {
+        return startSpan(methodName, table.getDatabaseName(), table.getTableName(), partitionName);
+    }
+
+    private Span startSpan(String methodName, String schemaName, String tableName, String partitionName)
+    {
+        return tracer.spanBuilder("Metadata." + methodName)
+                .startSpan()
+                .setAttribute(SCHEMA, schemaName)
+                .setAttribute(TABLE, tableName)
+                .setAttribute(PARTITION, partitionName);
+    }
+
+    @MustBeClosed
+    private static ScopedSpan scopedSpan(Span span)
+    {
+        return new ScopedSpan(span);
+    }
+
+    private static final class ScopedSpan
```

Comment:
Consider doing this similar to `io.trino.filesystem.tracing.Tracing.withTracing()` so that we can record the exception